### DISCUSSION TITLE
How to build Bevy apps in remote development environment?

### DISCUSSION THREAD

**Discussion by jsjoeio** [bevyengine/bevy#4953] [Q&A] [answered]
Hi! New to Bevy. I use a remote development environment (Ubuntu based Docker container). I connect via VS Code in the browser (code-server). When I try and run a very basic Bevy app, I get this error:

```rust
thread 'main' panicked at 'Failed to initialize any backend! Wayland status: "backend disabled" X11 status: 
LibraryOpenError(OpenError { kind: Library, detail: 
"opening library failed (libXcursor.so.1: cannot open 
shared object file: No such file or directory); opening 
library failed (libXcursor.so: cannot open shared object 
file: No such file or directory)" })', 
/home/coder/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.1/src/platform_impl/linux/mod.rs:619:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Is this possible or does Bevy only work with local development environments?

**Comment by alice-i-cecile** [comment]
You might be interested in https://github.com/bevyengine/bevy/pull/4700, which sets up remote development using Gitpod. I'm not immediately sure how adaptable it might be, but it's a place to look.

**Reply by jsjoeio** [reply]
Wow, that has to be the world's fastest reply! I'll take a look 👀

**Comment by jsjoeio** [comment, marked as answer]
Turns out, it's easier than you think. Bevy can be compiled for WebAssembly. Follow [this guide](https://bevy-cheatbook.github.io/platforms/wasm.html#quick-start) and it will run your game on a webserver.

**Comment by mockersf** [comment]
if your game can't work in wasm (need more shadows, need more perf, need multithread, ...), there are still options open to you:
* run your game on your server. you'll need a server with a GPU, and something like X11 forwarding or screen sharing
* cross compile your game on your server for your local setup, then download the binary and run it locally. you'll need a GPU locally in this case

**Reply by bjorn3** [reply]
I believe wayland + waypipe works better than X11 for forwarding programs using hardware acceleration I believe. Waypipe forwards the framebuffer as an (h.264?) compressed video.

**Reply by AgustinRamiroDiaz** [reply]
Related post to this idea: https://jasoncg.isthecode.ninja/2023/09/20/developing-gui-wayland-app-bevy-in-vscode-devcontainer/

**Comment by nomadshiba** [comment]
You can also run `cargo rustc` then run the built from host. So you only have development related things inside the container.
And can test the thing directly inside your host machine. I think this approach is more clean.
You can also use SSH which makes everything easier.

Here how I did it:
- First you gotta have SSH open on your host of course.
- Then inside `devcontainer.json` add these.
You will need these with in the container.
```json
"containerEnv": {
    "HOST_USER": "${localEnv:USER}",
    "HOST_WORKSPACE": "${localWorkspaceFolder}",
    "HOST_DISPLAY": "${localEnv:DISPLAY}",
    "HOST_XAUTHORITY": "${localEnv:XAUTHORITY}"
}
```

- And then inside your project create a file named `run` or any other name you want.
Inside the script we first run `cargo rustc` if it was successful, then we `ssh` connect to the host machine and run the thing.
```bash
#!/bin/bash

# Build the project
cargo rustc

# Check if the build was successful, otherwise exit
if [ $? -ne 0 ]; then
    exit 1
fi

# Get the host user and ip address
REMOTE_ADDRESS="${HOST_USER}@$(ip route | grep default | awk '{print $3}')"

# Check if SSH key pair exists, otherwise generate one
if [ ! -f ~/.ssh/id_rsa ]; then
  ssh-keygen -t rsa -N "" -f ~/.ssh/id_rsa
fi

# Copy the public key to the local host if not already done
ssh-copy-id -i ~/.ssh/id_rsa.pub ${REMOTE_ADDRESS} >/dev/null 2>&1

# Run the program on the local host
ssh ${REMOTE_ADDRESS} \
    "XAUTHORITY=${HOST_XAUTHORITY} DISPLAY=${HOST_DISPLAY} ${HOST_WORKSPACE}/target/debug/$(basename "${HOST_WORKSPACE}")"

```

If im missing something let me know, just started using rust and bevy today.
This was the best way to get started for me since i use dev containers within flatpak vscode and dont let development related things on my host system.

**Comment by cs50victor** [comment]
Leaving this here in hopes it helps someone in future : https://github.com/cs50victor/new_media/pull/18

**Comment by 1mamute** [comment]
I've been using [VSCode's devcontainer](https://code.visualstudio.com/docs/devcontainers/containers) to develop bevy for a few days. There were a lot of fiddling around until I've managed to make it work: mounting GPU, X11 and sound sockets, installing NVIDIA containers toolkit on host machine, installing cuda toolkit and drivers inside container.

Relevant informations about my setup: Ubuntu 22.04, NVIDIA GPU with proprietary drivers, [system user mapped to docker group](https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user)
The exact requirements may depend on your setup, but it may serve as a starting point:

1. Install and configure [NVIDIA's Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html) on your host machine: required to mount the GPU into containers
```bash
curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
&& curl -s -L https://nvidia.github.io/libnvidia-container/stable/deb/nvidia-container-toolkit.list | \
sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
sudo apt-get install -y nvidia-container-toolkit
sudo nvidia-ctk runtime configure --runtime=docker
```
2. Enable X11 forwarding using the command `xhost +`, remember to disable later `xhost -`
2. Mount the GPU, X11, sound sockets and set the DISPLAY environment variable inside the container:
```json
// .devcontainer/devcontainer.json
{
	"name": "Rust",
	"build": {
		"dockerfile": "Dockerfile"
	},

	"runArgs": [
		// Map sound devices (doesn't work on Windows, see https://github.com/microsoft/WSL/issues/689)
                "--device=/dev/snd:/dev/snd",
                "--device=/dev/dri:/dev/dri",
		"-e", "DISPLAY=${env:DISPLAY}", // Set the display for X11
                "-v", "/tmp/.X11-unix:/tmp/.X11-unix", // Mount the X11 socket
		"--gpus", "all", // Mount the GPU
                "--privileged" // It may or not be necessary to run the container in privileged mode
	]
}
```
3. Install required drivers and tools inside the container, for me, the following were required
```dockerfile
# .devcontainer/Dockerfile
FROM mcr.microsoft.com/devcontainers/rust:1-1-bullseye

# Install cuda-toolkit
RUN distribution=$(. /etc/os-release;echo $ID$VERSION_ID | sed -e 's/\.//g') && \
    wget https://developer.download.nvidia.com/compute/cuda/repos/$distribution/x86_64/cuda-keyring_1.0-1_all.deb && \
    dpkg -i cuda-keyring_1.0-1_all.deb && \
    export DEBIAN_FRONTEND=noninteractive && \
    apt-get -y update && \
    apt-get -y install cuda-toolkit && \
    rm cuda*

# Install additional packages.
RUN export DEBIAN_FRONTEND=noninteractive && apt-get -y upgrade \
    && apt-get -y install --no-install-recommends \
    g++ \
    pkg-config \
    libx11-dev \
    libasound2-dev \
    libudev-dev \
    libxkbcommon-x11-0 \
    libxcursor-dev \
    libxrandr-dev \
    libxi-dev \
    mesa-vulkan-drivers \
    lld \
    clang

# Cargo tools for nightly must be installed with vscode user
USER vscode
RUN rustup toolchain install nightly && \
    rustup component add rustfmt --toolchain nightly-x86_64-unknown-linux-gnu && \
    rustup component add clippy --toolchain nightly-x86_64-unknown-linux-gnu
```

Good luck!
